### PR TITLE
Additional flags

### LINKED ISSUE #40: cdncheck cli support + update
Continuation of https://github.com/projectdiscovery/cdncheck/pull/37

- [x] CLI flag add / update
- [x] JSON output

#### CLI option:

```console
cdncheck is a tool for identifying the technology associated with ip network addresses.

Usage:
  ./cdncheck [flags]

Flags:
INPUT:
   -i, -inputs string[]  input ip / cidr to process
   -l, -list string      file input with list of ip / cidr to process

DETECTION:
   -cdn              display cdn ip
   -cloud            display cloud ip
   -waf              display waf ip

MATCHERS:
   -mcdn, -match-cdn string[]         match host with specified cdn provider (fastly, incapsula)
   -mcloud, -match-cloud string[]     match host with specified cloud provider (fastly, incapsula)
   -mwaf, -match-waf string[]         match host with specified waf provider (fastly, incapsula)

FILTER:
   -fcdn, -filter-cdn string[]        filter host with specified cdn provider (fastly, incapsula)
   -mcloud, -filter-cdn string[]      filter host with specified cloud provider (fastly, incapsula)
   -mwaf, -filter-cdn string[]        filter host with specified waf provider (fastly, incapsula)
   -e, -exclude      		      exclude detected ip from output

OUTPUT:
   -resp             display technology name in cli output
   -j, -json         write output in json format
   -o, -output       write output in plain format
   -version          display version of the project
```

#### JSON output:

```json
{
  "timestamp": "2022-08-02T03:00:17.698783+05:30",
  "ip": "104.16.51.111",
  "cdn": true,
  "cdn_name": "cloudflare"
}
```

#### CLI examples:

```bash
$ cat ips

52.60.165.183
3.98.63.202
52.60.160.16
185.199.109.153
104.16.51.111
54.192.171.16
54.192.171.2
54.192.171.22
```

```bash
$ cat ips | cdncheck

104.16.51.111 
54.192.171.16
54.192.171.2
54.192.171.22
```

```bash
$ cat ips | cdncheck -resp

104.16.51.111 [waf] [cloudflare] 
54.192.171.16 [waf] [cloudflare] 
54.192.171.2 [cdn] [cloudfront] 
54.192.171.22 [cdn] [cloudfront] 
```

```bash
$ cat ips | cdncheck -resp -waf

104.16.51.111 [waf] [cloudflare] 
54.192.171.16 [waf] [cloudflare]
```

```bash
$ cat ips | cdncheck -e

52.60.165.183
3.98.63.202
52.60.160.16
185.199.109.153
```

```bash
$ cat ips | cdncheck -e -waf

52.60.165.183
3.98.63.202
52.60.160.16
185.199.109.153
54.192.171.2
54.192.171.22
```
Question: `-e` should already exclude `-waf`?

### PR DESCRIPTION
**Pull request by LuitelSamikshya**
closes #40 
```console
MacBook-Pro:cdncheck user$ cat list.txt
52.60.165.183
3.98.63.202
52.60.160.16
185.199.109.153
54.192.171.16
54.192.171.2
54.192.171.22
104.16.51.111
MacBook-Pro:cdncheck user$ go build && ./cdncheck -l list.txt -resp
52.60.165.183 [cloud] [aws]
3.98.63.202 [cloud] [aws]
52.60.160.16 [cloud] [aws]
54.192.171.16 [cdn] [cloudfront]
54.192.171.2 [cdn] [cloudfront]
54.192.171.22 [cdn] [cloudfront]
104.16.51.111 [waf] [cloudflare]
MacBook-Pro:cdncheck user$ go build && ./cdncheck -l list.txt -e
185.199.109.153
MacBook-Pro:cdncheck user$ go build && ./cdncheck -l list.txt -resp -cdn
54.192.171.16 [cdn] [cloudfront]
54.192.171.2 [cdn] [cloudfront]
54.192.171.22 [cdn] [cloudfront]
MacBook-Pro:cdncheck user$ go build && ./cdncheck -l list.txt -resp -mcdn cloudfront
54.192.171.16 [cdn] [cloudfront]
54.192.171.2 [cdn] [cloudfront]
54.192.171.22 [cdn] [cloudfront]
MacBook-Pro:cdncheck user$ go build && ./cdncheck -l list.txt -resp -fcdn cloudfront
52.60.165.183 [cloud] [aws]
3.98.63.202 [cloud] [aws]
52.60.160.16 [cloud] [aws]
104.16.51.111 [waf] [cloudflare]
MacBook-Pro:cdncheck user$ go build && ./cdncheck -l list.txt -resp -waf
104.16.51.111 [waf] [cloudflare]
MacBook-Pro:cdncheck user$ go build && ./cdncheck -l list.txt -resp -waf -mwaf cloudflare
104.16.51.111 [waf] [cloudflare]
MacBook-Pro:cdncheck user$ go build && ./cdncheck -l list.txt -resp -waf -fwaf cloudflare
MacBook-Pro:cdncheck user$ go build && ./cdncheck -l list.txt -resp -cloud
52.60.165.183 [cloud] [aws]
3.98.63.202 [cloud] [aws]
52.60.160.16 [cloud] [aws]
MacBook-Pro:cdncheck user$ go build && ./cdncheck -l list.txt -resp -mcloud aws
52.60.165.183 [cloud] [aws]
3.98.63.202 [cloud] [aws]
52.60.160.16 [cloud] [aws]
MacBook-Pro:cdncheck user$ go build && ./cdncheck -l list.txt -resp -fcloud aws
54.192.171.16 [cdn] [cloudfront]
54.192.171.2 [cdn] [cloudfront]
54.192.171.22 [cdn] [cloudfront]
104.16.51.111 [waf] [cloudflare]

```

### CODE REVIEW

**Review by ehsandeep — CHANGES_REQUESTED** [review]
@LuitelSamikshya In the JSON output, `cdn` and `cdn_name` is used for `cloud` and `waf` as well.

```
echo 54.237.133.81 | ./cdncheck -resp
54.237.133.81 [cloud] [aws]
```

```
$echo 54.237.133.81 | ./cdncheck -resp -json | jq .

{
  "timestamp": "2022-08-20T00:13:24.499455+04:00",
  "ip": "54.237.133.81",
  "cdn": true,
  "cdn_name": "aws"
}
```

File: internal/runner/options.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+	flagSet.CreateGroup("output", "Output",
+		flagSet.BoolVarP(&opts.response, "resp", "", false, "display technology name in cli output"),
+		flagSet.StringVarP(&opts.output, "output", "o", "", "write output in plain format to file"),
+		flagSet.BoolVarP(&opts.version, "version", "", false, "display version of the project"),
+		flagSet.BoolVarP(&opts.json, "json", "", false, "write output in json format to file"),
+	)
+
+	flagSet.CreateGroup("matchers", "Matchers",
+		flagSet.StringSliceVarP(&opts.matchCdn, "match-cdn", "mcdn", nil, "match host with specified cdn provider (fastly, incapsula)", goflags.CommaSeparatedStringSliceOptions),
```
Comment (ehsandeep) [inline internal/runner/options.go:83]: These values need to be dynamic, similar to https://github.com/projectdiscovery/httpx/blob/f2cd5b835348c5563c4c018dc2ed414d82ccf249/runner/options.go#L309, currently `fastly, incapsula` is hard coded for all options.

Comment (LuitelSamikshya) [reply internal/runner/options.go]: Done

**Review by ehsandeep — APPROVED** [review]

File: cdncheck.go [hunk 2]
```diff
@@ -53,3 +53,28 @@ func (c *Client) Check(ip net.IP) (matched bool, value string, itemType string,
 	}
 	return false, "", "", err
 }
+
+// GetDefaultProviders exports default providers
+func GetDefaultCdnProviders() []string {
+	var provider []string
+	for k := range generatedData.CDN {
```
Comment (Mzack9999) [inline cdncheck.go:60]: We can package the map keys extraction into a helper function (e.g. `mapKeys`).
Another alternative can be to bump the required go version to 1.18 and use the new function `maps.Keys` from the standard library (ref https://pkg.go.dev/golang.org/x/exp/maps#Keys)

File: cdncheck.go [hunk 3]
```diff
@@ -53,3 +53,28 @@ func (c *Client) Check(ip net.IP) (matched bool, value string, itemType string,
...
+		provider = append(provider, k)
+	}
+	return provider
+}
+
+func GetDefaultCloudProviders() []string {
+	var provider []string
+	for k := range generatedData.Cloud {
```
Comment (Mzack9999) [inline cdncheck.go:68]: see my previous comment

File: cdncheck.go [hunk 4]
```diff
@@ -53,3 +53,28 @@ func (c *Client) Check(ip net.IP) (matched bool, value string, itemType string,
...
+		provider = append(provider, k)
+	}
+	return provider
+}
+
+func GetDefaultWafProviders() []string {
+	var provider []string
+	for k := range generatedData.WAF {
```
Comment (Mzack9999) [inline cdncheck.go:76]: see my previous comment

**Review by Mzack9999 — CHANGES_REQUESTED** [review]

File: cdncheck.go (hunk 2 above)
Comment (LuitelSamikshya) [reply cdncheck.go:64]: Used helper function

File: cdncheck.go (hunk 3 above)
Comment (LuitelSamikshya) [reply cdncheck.go:69]: Done

File: cdncheck.go (hunk 4 above)
Comment (LuitelSamikshya) [reply cdncheck.go:74]: Done

File: cdncheck.go [hunk 5]
```diff
@@ -53,3 +56,29 @@ func (c *Client) Check(ip net.IP) (matched bool, value string, itemType string,
...
+	}
+	providers["cloud"] = provider
+	provider = nil
+	for k := range generatedData.WAF {
+		provider = append(provider, k)
+	}
+	providers["waf"] = provider
+	return providers
+}
+func MapKeys(m map[string][]string, key string) string {
```
Comment (Mzack9999) [inline cdncheck.go:80]: The function should be only internal:
```go
func mapKeys(m map[string][]string) []string {
	var keys []string
	for key := range m {
		keys = append(keys, m[key]...)
	}
	return keys
}
```

File: cdncheck.go [hunk 6]
```diff
@@ -53,3 +56,29 @@ func (c *Client) Check(ip net.IP) (matched bool, value string, itemType string,
 	}
 	return false, "", "", err
 }
+
+// GetDefaultProviders exports default providers
+func GetDefaultProviders() map[string][]string {
```
Comment (Mzack9999) [inline cdncheck.go:61]: Why just not exporting them like this:
```go
var (
	DefaultCDNProviders  = mapKeys(generatedData.CDN)
	DefaultWarProviders  = mapKeys(generatedData.WAF)
	DefaultCloudProviders = mapKeys(generatedData.Cloud)
)
```

**Review by Mzack9999 — CHANGES_REQUESTED** [review]

File: cdncheck.go (hunk 5 above)
Comment (LuitelSamikshya) [reply cdncheck.go]: Done

File: cdncheck.go (hunk 6 above)
Comment (LuitelSamikshya) [reply cdncheck.go]: Sure

**Review by Mzack9999 — APPROVED** [review]

**Review by ehsandeep — APPROVED** [review]

**Pull request merged**